Sources/WebSocketKit/PMCE.swift

```diff
@@ -0,0 +1,753 @@
+                                                   eventLoop: channel.eventLoop)
+            logger.trace("compressor \(zscConf)")
+            logger.trace("decompressor \(zsdConf)")
+            
+        case .client:
+            logger.trace("client")
+
+            let winSize = PMCE.sizeFor(bits: clientConfig.deflateConfig.agreedParams.maxWindowBits ?? 15)
+            logger.trace("window size: \(winSize)")
+
+            let zccConf = ZlibConfiguration(windowSize: winSize,
+                                            compressionLevel: clientConfig.deflateConfig.zlibConfig.compressionLevel,
+                                            memoryLevel: clientConfig.deflateConfig.zlibConfig.memLevel,
+                                          strategy: .huffmanOnly)
+            
+            let zcdConf = ZlibConfiguration(windowSize: winSize,
+                                            compressionLevel: clientConfig.deflateConfig.zlibConfig.compressionLevel,
+                                            memoryLevel: clientConfig.deflateConfig.zlibConfig.memLevel,
+                                          strategy: .huffmanOnly)
+
+            self.compressorBox = NIOLoopBoundBox(CompressionAlgorithm.deflate(configuration: zccConf).compressor,
+                                                 eventLoop: channel.eventLoop)
+            
+            self.decompressorBox = NIOLoopBoundBox( CompressionAlgorithm.deflate(configuration: zcdConf).decompressor,
+                                             eventLoop: channel.eventLoop)
+            
+            logger.trace("compressor \(zccConf)")
+            logger.trace("decompressor \(zcdConf)")
+        }
+        startStreams()
+    }
+    
+    ///  Starts compress-nio streams for DEFLATE support.
+    /// - returns: Void
+    public func startStreams() {
+        do {
+            try compressorBox.value?.startStream()
+        }
+        catch {
+            logger.error("error starting compressor stream : \(error)")
+        }
+        do {
+            try decompressorBox.value?.startStream()
+        }
+        catch {
+            logger.error("error starting decompressor stream : \(error)")
+        }
+    }
+    
+    ///  Stops compress-nio streams for DEFLATE support.
+    /// - returns: Void
+    public func stopStreams() {
+        do {
+            logger.debug("PMCE: stopping compressor stream...")
+            try compressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+        do {
+            logger.debug("PMCE: stopping decompressor stream...")
+            try decompressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+
+    }
+
+    /// websocket send calls this to compress.
+    public func compressed(_ buffer: ByteBuffer,
```

Comment:
Why public?

Reply:
same as above had considered extracting it and since something external to PMCE class it seemed logical to make it public.